### PR TITLE
[Debugger Expression Evaluation] Fix support for 'array' and 'value' langlib invocations

### PR DESCRIPTION
**Pull request by NipunaRanasinghe**
## Purpose
Fixes https://github.com/ballerina-platform/ballerina-lang/issues/29563 - Getting `Failed to load lang library: 'lang.object, lang.value'`, when evaluating expressions.

## Approach
The debugger expression evaluation engine is now using the syntax API of the lang lib packages (which are resolved using the package cache), to resolve the target function definition (in order to validate user arguments).

## Remarks
With this new approach, the requirement (https://github.com/ballerina-platform/ballerina-lang/issues/27520) from the semantic API will become invalid.

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [x] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [x] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/evaluation/engine/FunctionInvocationExpressionEvaluator.java [hunk 1]
```diff
@@ -91,8 +91,7 @@ public BExpressionValue evaluate() throws EvaluationException {
         SemanticModel semanticContext = context.getDebugCompiler().getSemanticInfo();
         List<Symbol> functionMatches = semanticContext.moduleSymbols()
                 .stream()
-                .filter(symbol -> symbol.kind() == SymbolKind.FUNCTION
-                        && modifyName(symbol.getName().get()).equals(functionName))
+                .filter(symbol -> symbol.kind() == FUNCTION && modifyName(symbol.getName().get()).equals(functionName))
```
Comment (praveennadarajah) [inline misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/evaluation/engine/FunctionInvocationExpressionEvaluator.java:94]: Will `symbol.getName()` always present? Shall we add `isPresent` check for `symbol.getName()`?
```suggestion
                .filter(symbol -> symbol.kind() == FUNCTION && symbol.getName().isPresent() && modifyName(symbol.getName().get()).equals(functionName))
```

File: misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/evaluation/engine/FunctionNodeFinder.java [hunk 2]
```diff
@@ -33,26 +33,32 @@
...
                 break;
             }
-            searchIn(module.document(documentId));
+            searchInFile(module.document(documentId));
         }
         return Optional.ofNullable(result);
     }
 
-    public Optional<FunctionDefinitionNode> searchIn(Document document) {
+    public Optional<FunctionDefinitionNode> searchInFile(Document document) {
```
Comment (praveennadarajah) [inline misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/evaluation/engine/FunctionNodeFinder.java:61]: Are we using the return value of `searchInFile` method ? If not, shall we make this method `void`?

File: misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/evaluation/engine/InvocationArgProcessor.java [hunk 3]
```diff
@@ -80,7 +88,15 @@
...
-                argValues.put(parameterName, arg.getValue().evaluate().getJdiValue());
+                Value argValue = arg.getValue().evaluate().getJdiValue();
+                // For the ballerina function parameters with type "any", the generated runtime method will only accept
+                // the args which are subtypes of "java.lang.Object". Therefore all the primitive typed arguments
+                // must be converted into their wrapper implementations (i.e. 'int' -> 'Integer'), before passing
+                // into the method.
+                if (params.get(i).typeDescriptor().typeKind().getName().equals(BVariableType.ANY.getString())) {
+                    argValue = getValueAsObject(context, argValue);
+                }
+                argValues.put(parameterName, argValue);
```
Comment (praveennadarajah) [inline misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/evaluation/engine/InvocationArgProcessor.java:99]: Shall we add this to a util method and use it here?

File: misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/evaluation/engine/InvocationArgProcessor.java [hunk 4]
```diff
@@ -94,7 +110,15 @@
...
-                argValues.put(argName, arg.getValue().evaluate().getJdiValue());
+                Value argValue = arg.getValue().evaluate().getJdiValue();
+                // For the ballerina function parameters with type "any", the generated runtime method will only accept
+                // the args which are subtypes of "java.lang.Object". Therefore all the primitive typed arguments
+                // must be converted into their wrapper implementations (i.e. 'int' -> 'Integer'), before passing
+                // into the method.
+                if (params.get(i).typeDescriptor().typeKind().getName().equals(BVariableType.ANY.getString())) {
+                    argValue = getValueAsObject(context, argValue);
+                }
+                argValues.put(argName, argValue);
```
Comment (praveennadarajah) [inline misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/evaluation/engine/InvocationArgProcessor.java:121]: Same as [here](https://github.com/ballerina-platform/ballerina-lang/pull/30205/files#r621760910)

File: misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/evaluation/engine/InvocationArgProcessor.java [hunk 5]
```diff
@@ -134,6 +158,136 @@
...
+            Map.Entry<String, Evaluator> arg = argEvaluators.get(i);
+            ArgType argType = getArgType(arg);
+            if (argType == ArgType.POSITIONAL) {
+                if (namedArgsFound) {
+                    throw new EvaluationException(String.format(EvaluationExceptionKind.CUSTOM_ERROR.getString(),
+                            "positional args are not allowed after named args."));
+                } else if (restArgsFound) {
+                    throw new EvaluationException(String.format(EvaluationExceptionKind.CUSTOM_ERROR.getString(),
+                            "positional args are not allowed after rest args."));
+                }
```
Comment (praveennadarajah) [inline misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/evaluation/engine/InvocationArgProcessor.java:187]: Do we need a `else if` here? Instead shall we add
```
if (namedArgsFound || restArgsFound) {
    throw new EvaluationException()
}
```

File: misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/evaluation/engine/InvocationArgProcessor.java [hunk 6]
```diff
@@ -134,6 +158,136 @@
...
+                String parameterName = getParameterName(params.get(i));
+                Value argValue = arg.getValue().evaluate().getJdiValue();
+                // For the ballerina function parameters with type "any", the generated runtime method will only accept
+                // the args which are subtypes of "java.lang.Object". Therefore all the primitive typed arguments
+                // must be converted into their wrapper implementations (i.e. 'int' -> 'Integer'), before passing
+                // into the method.
+                if (getParameterTypeName(params.get(i)).equals(BVariableType.ANY.getString())) {
+                    argValue = getValueAsObject(context, argValue);
+                }
+                argValues.put(parameterName, argValue);
```
Comment (praveennadarajah) [inline misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/evaluation/engine/InvocationArgProcessor.java:203]: Same as [here](https://github.com/ballerina-platform/ballerina-lang/pull/30205/files#r621760910)

File: misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/evaluation/engine/InvocationArgProcessor.java [hunk 7]
```diff
@@ -134,6 +158,136 @@
...
+                namedArgsFound = true;
+                Value argValue = arg.getValue().evaluate().getJdiValue();
+                // For the ballerina function parameters with type "any", the generated runtime method will only accept
+                // the args which are subtypes of "java.lang.Object". Therefore all the primitive typed arguments
+                // must be converted into their wrapper implementations (i.e. 'int' -> 'Integer'), before passing
+                // into the method.
+                if (getParameterTypeName(params.get(i)).equals(BVariableType.ANY.getString())) {
+                    argValue = getValueAsObject(context, argValue);
+                }
+                argValues.put(argName, argValue);
```
Comment (praveennadarajah) [inline misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/evaluation/engine/InvocationArgProcessor.java:225]: Same as [here](https://github.com/ballerina-platform/ballerina-lang/pull/30205/files#r621760910)

File: misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/evaluation/engine/InvocationArgProcessor.java [hunk 8]
```diff
@@ -134,6 +158,136 @@
...
+    }
+
+    private static String getParameterName(ParameterNode parameterNode) {
+        switch (parameterNode.kind()) {
+            case REQUIRED_PARAM:
+                return ((RequiredParameterNode) parameterNode).paramName().get().text();
+            case DEFAULTABLE_PARAM:
+                return ((DefaultableParameterNode) parameterNode).paramName().get().text();
+            case REST_PARAM:
+                return ((RestParameterNode) parameterNode).paramName().get().text();
```
Comment (praveennadarajah) [inline misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/evaluation/engine/InvocationArgProcessor.java:272]: Will `parameterNode).paramName()` always present? Shall we add `isPresent` check for `parameterNode).paramName()` ?

File: misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/evaluation/engine/FunctionInvocationExpressionEvaluator.java (hunk 1 above)
Comment (NipunaRanasinghe) [reply misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/evaluation/engine/FunctionInvocationExpressionEvaluator.java]: +1. Added in 8a8a248

File: misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/evaluation/engine/FunctionNodeFinder.java (hunk 2 above)
Comment (NipunaRanasinghe) [reply misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/evaluation/engine/FunctionNodeFinder.java:64]: Refactored to use the return value in 8a8a248

File: misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/evaluation/engine/InvocationArgProcessor.java (hunk 3 above)
Comment (NipunaRanasinghe) [reply misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/evaluation/engine/InvocationArgProcessor.java:101]: This is already a util method (`getValueAsObject(context, argValue)`);

File: misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/evaluation/engine/InvocationArgProcessor.java (hunk 4 above)
Comment (NipunaRanasinghe) [reply misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/evaluation/engine/InvocationArgProcessor.java:123]: https://github.com/ballerina-platform/ballerina-lang/pull/30205#discussion_r621798365

File: misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/evaluation/engine/InvocationArgProcessor.java (hunk 5 above)
Comment (NipunaRanasinghe) [reply misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/evaluation/engine/InvocationArgProcessor.java:189]: We cannot, as the Error messages are different.

File: misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/evaluation/engine/InvocationArgProcessor.java (hunk 6 above)
Comment (NipunaRanasinghe) [reply misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/evaluation/engine/InvocationArgProcessor.java:205]: https://github.com/ballerina-platform/ballerina-lang/pull/30205#discussion_r621798365

File: misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/evaluation/engine/InvocationArgProcessor.java (hunk 7 above)
Comment (NipunaRanasinghe) [reply misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/evaluation/engine/InvocationArgProcessor.java:227]: https://github.com/ballerina-platform/ballerina-lang/pull/30205#discussion_r621798365

File: misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/evaluation/engine/InvocationArgProcessor.java (hunk 8 above)
Comment (NipunaRanasinghe) [reply misc/debug-adapter/modules/debug-adapter-core/src/main/java/org/ballerinalang/debugadapter/evaluation/engine/InvocationArgProcessor.java]: +1, added in 8a8a248

**Review by praveennadarajah — APPROVED** [review]

**Pull request merged**